src/main/java/spliterators/part1/exercise/RectangleSpliterator.java

```diff
@@ -40,7 +49,29 @@ public long estimateSize() {
 
     @Override
     public boolean tryAdvance(IntConsumer action) {
-        // TODO
-        throw new UnsupportedOperationException();
+        if (startOuterInclusive >= endOuterExclusive) {
+            return false;
+        }
+        if (startInnerInclusive >= innerLength) {
+            if (startOuterInclusive < endOuterExclusive -1) {

```

Comment:
You already have a similar `if` block. You can refactor this core to reduce `if`s count.